File: app/assets/javascripts/angular/controllers/work-packages-controller.js
//-- copyright
// OpenProject is a project management system.
// Copyright (C) 2012-2014 the OpenProject Foundation (OPF)
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License version 3.
//
// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
// Copyright (C) 2006-2013 Jean-Philippe Lang
// Copyright (C) 2010-2013 the ChiliProject Team
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// See doc/COPYRIGHT.rdoc for more details.
//++

angular.module('openproject.workPackages.controllers')

.controller('WorkPackagesController', [
    '$scope',
    '$rootScope',
    '$q',
    '$window',
    '$location',
    'I18n',
    'ProjectService',
    'WorkPackagesTableService',
    'WorkPackageService',
    'QueryService',
    'PaginationService',
    'WorkPackageLoadingHelper',
    'INITIALLY_SELECTED_COLUMNS',
    'OPERATORS_AND_LABELS_BY_FILTER_TYPE',
    function($scope, $rootScope, $q, $window, $location, I18n, ProjectService,
      WorkPackagesTableService,
      WorkPackageService, QueryService, PaginationService,
      WorkPackageLoadingHelper, INITIALLY_SELECTED_COLUMNS,
      OPERATORS_AND_LABELS_BY_FILTER_TYPE) {

  // Setup

  function initialSetup() {
    setupPageParamsFromUrl($window.location);
    initProject();

    $scope.selectedTitle = I18n.t('js.toolbar.unselected_title');
    $scope.operatorsAndLabelsByFilterType = OPERATORS_AND_LABELS_BY_FILTER_TYPE;
    $scope.loading = false;
    $scope.disableFilters = false;

    var getWorkPackages, params;
    if($location.search()['c[]']){
      getWorkPackages = WorkPackageService.getWorkPackagesFromUrlQueryParams;
      params = [$scope.projectIdentifier, $location];
    } else {
      getWorkPackages = WorkPackageService.getWorkPackagesByQueryId;
      params = [$scope.projectIdentifier, $scope.query_id];
    }

    $scope.withLoading(getWorkPackages, params)
      .then(setupPage);
  }

  function setupPageParamsFromUrl(location) {
    var normalisedPath = location.pathname.replace($window.appBasePath, '');

    $scope.projectIdentifier = normalisedPath.split('/')[2];
    $scope.$root.projectIdentifier = $scope.projectIdentifier; // temporary solution to share project identifier

    var regexp = /query_id=(\d+)/g;
    var match = regexp.exec(location.search);
    if(match) $scope.query_id = match[1];
  }


  function initProject() {
    if ($scope.projectIdentifier) {
      ProjectService.getProject($scope.projectIdentifier).then(function(project) {
        $scope.project  = project;
        $scope.projects = [ $scope.project ];
        $scope.availableTypes = $scope.project.embedded.types;
      });
    } else {
      ProjectService.getProjects().then(function(projects) {
        var allTypes, availableTypes;

        $scope.projects = projects;
        allTypes = projects.map(function(project) {
          return project.embedded.types;
        }).reduce(function(a, b) {
          return a.concat(b);
        }, []);

        $scope.availableTypes = allTypes; // TODO remove duplicates
      });
    }
  }

  function setupPage(json) {
    initQuery(json.meta);
    setupWorkPackagesTable(json);

    return $q.all([
      initAvailableColumns(),
      initAvailableQueries()
    ]);
  }

  function initQuery(metaData) {
    var queryData = metaData.query,
        columnData = metaData.columns;

    $scope.query = QueryService.getQuery() || QueryService.initQuery($scope.query_id, queryData, columnData, afterQuerySetupCallback);
  }

  function afterQuerySetupCallback(query) {
    $scope.showFiltersOptions = query.filters.length > 0;
    $scope.updateBackUrl();
  }

  function setupWorkPackagesTable(json) {
    var meta = json.meta,
        workPackages = json.work_packages,
        bulkLinks = json._bulk_links;

    // register data

    // table data
    WorkPackagesTableService.setColumns($scope.query.columns);
    WorkPackagesTableService.addColumnMetaData(meta);
    WorkPackagesTableService.setGroupBy($scope.query.groupBy);
    WorkPackagesTableService.buildRows(workPackages, $scope.query.groupBy);
    WorkPackagesTableService.setBulkLinks(bulkLinks);

    // query data
    QueryService.setTotalEntries(meta.total_entries);

    // pagination data
    PaginationService.setPerPageOptions(meta.per_page_options);
    PaginationService.setPerPage(meta.per_page);
    PaginationService.setPage(meta.page);


    // yield updatable data to scope
    $scope.columns = $scope.query.columns;
    $scope.rows = WorkPackagesTableService.getRows();
    $scope.groupableColumns = WorkPackagesTableService.getGroupableColumns();
    $scope.workPackageCountByGroup = meta.work_package_count_by_group;
    $scope.totalEntries = QueryService.getTotalEntries();

    // back url
    $scope.updateBackUrl();
  }

  function initAvailableColumns() {
    return QueryService.loadAvailableUnusedColumns($scope.projectIdentifier)
      .then(function(data){
        $scope.availableUnusedColumns = data;
      });
  }

  function initAvailableQueries() {
    QueryService.loadAvailableGroupedQueries($scope.projectIdentifier);

    $scope.availableOptions = QueryService.getAvailableOptions(); // maybe generalize this approach
  }

  $scope.$watch('availableOptions.availableGroupedQueries', function(availableQueries) {
    if (availableQueries) {
      $scope.groups = [{ name: 'CUSTOM QUERIES', models: availableQueries['user_queries']},
                       { name: 'GLOBAL QUERIES', models: availableQueries['queries']}];
    }
  });

  // Updates

  $scope.reloadQuery = function(queryId) {
    QueryService.resetQuery();
    $scope.query_id = queryId;

    $scope.withLoading(WorkPackageService.getWorkPackagesByQueryId, [$scope.projectIdentifier, $scope.query_id])
      .then(setupPage);
  };

  $scope.updateBackUrl = function(){
    // Easier than trying to extract it from $location
    var relativeUrl = "/work_packages";
    if ($scope.projectIdentifier){
      relativeUrl = "/projects/" + $scope.projectIdentifier + relativeUrl;
    }

    if($scope.query){
      relativeUrl = relativeUrl + "#?" + $scope.query.getQueryString();
    }

    $scope.backUrl = relativeUrl;
  };

  $scope.updateResults = function() {
    $scope.$broadcast('openproject.workPackages.updateResults');

    return $scope.withLoading(WorkPackageService.getWorkPackages, [$scope.projectIdentifier, $scope.query, PaginationService.getPaginationOptions()])
      .then(setupWorkPackagesTable);
  };

  // More

  function serviceErrorHandler(data) {
    // TODO RS: This is where we'd want to put an error message on the dom
    $scope.isLoading = false;
  }

  $scope.withLoading = function(callback, params){
    return WorkPackageLoadingHelper.withLoading($scope, callback, params, serviceErrorHandler);
  };

  // Go

  initialSetup();

  // Note: I know we don't want watchers on the controller but I want all the toolbar directives to have restricted scopes. Thoughts welcome.
  $scope.$watch('query.name', function(newValue, oldValue){
    if(newValue != oldValue && $scope.query.hasName()){
      $scope.selectedTitle = newValue;
    }
  });

  $rootScope.$on('queryResetRequired', function(event, message) {
    $scope.query_id = null;
    initialSetup();
  });

}]);

File: app/controllers/api/experimental/work_packages_controller.rb
#-- copyright
# OpenProject is a project management system.
# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 3.
#
# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
# Copyright (C) 2006-2013 Jean-Philippe Lang
# Copyright (C) 2010-2013 the ChiliProject Team
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# See doc/COPYRIGHT.rdoc for more details.
#++

module Api
  module Experimental
    class WorkPackagesController < ApplicationController

      include ApiController
      include ::Api::Experimental::Concerns::GrapeRouting
      include ::Api::Experimental::Concerns::ColumnData
      include ::Api::Experimental::Concerns::QueryLoading
      include ::Api::Experimental::Concerns::V3Naming

      include PaginationHelper
      include QueriesHelper
      include SortHelper
      include ExtendedHTTP

      before_filter :find_optional_project,
                    :v3_params_as_internal,
                    :load_query

      def index
        @display_meta = true
        @work_packages_meta_data = {
          query:                        query_as_json(@query, User.current),
          columns:                      get_columns_for_json(@query.columns),
          groupable_columns:            get_columns_for_json(@query.groupable_columns),
          page:                         page_param,
          per_page:                     per_page_param,
          per_page_options:             Setting.per_page_options_array,
          export_formats:               export_formats,
          _links:                       work_packages_links
        }
      end

      private

      def load_query
        @query ||= init_query
      rescue ActiveRecord::RecordNotFound
        render_404
      end

      def work_packages_links
        links = {}
        links[:create] = api_experimental_work_packages_path(@project) if User.current.allowed_to?(:add_work_packages, @project)
        links[:export] = api_experimental_work_packages_path(@project) if User.current.allowed_to?(:export_work_packages, @project, global: @project.nil?)
        links
      end

      def query_as_json(query, user)
        json_query = query.as_json(except: :filters, include: :filters, methods: [:starred])
        # prefer using the identifier throughout the frontend so that we can
        # cache the requests more efficiently
        json_query["project_id"] = @project.identifier if @project

        json_query[:_links] = allowed_links_on_query(query, user)

        json_query_as_v3(json_query)
      end

      def export_formats
        [
          { identifier: 'atom', format: 'atom', label_locale: 'label_format_atom' },
          { identifier: 'pdf',  format: 'pdf', label_locale: 'label_format_pdf' },
          {
            identifier: 'pdf-descr',  format: 'pdf',
            label_locale: 'label_format_pdf_with_descriptions', flags: ['show_descriptions']
          },
          { identifier: 'csv', format: 'csv', label_locale: 'label_format_csv' }
        ]
      end
    end
  end
end
